File: src/shared/engine/globalActions.ts
import type { GameState, GamePhase, BoardType, Position, Move } from '../types/game';
import { positionToString } from '../types/game';
import {
  MovementBoardView,
  hasAnyLegalMoveOrCaptureFromOnBoard,
  computeProgressSnapshot,
} from './core';
import {
  enumeratePlacementPositions,
  evaluateSkipPlacementEligibility,
  SkipPlacementEligibilityResult,
} from './aggregates/PlacementAggregate';
import { enumerateProcessLineMoves } from './aggregates/LineAggregate';
import {
  enumerateProcessTerritoryRegionMoves,
  enumerateTerritoryEliminationMoves,
} from './aggregates/TerritoryAggregate';
import { applyEliminateRingsFromStackDecision } from './territoryDecisionHelpers';
import { isValidPosition } from './validators/utils';

/**
 * Summary of the global legal action surface G(state, P) for a player.
 *
 * This is a thin TS mirror of the RR-CANON R200–R203 definitions used by the
 * invariants framework:
 *
 * - hasTurnMaterial          – player is not fully eliminated (R201).
 * - hasGlobalPlacementAction – at least one legal ring placement exists,
 *                              ignoring currentPhase (R200).
 * - hasPhaseLocalInteractiveMove – at least one interactive move is legal in
 *                                  the current phase (placement/skip, movement/
 *                                  capture, line or territory decisions) (R200).
 * - hasForcedEliminationAction – R205/R072/R100 preconditions hold: player is
 *                                blocked with stacks, no placements or
 *                                movement/capture exist, and host-level forced
 *                                elimination is available.
 */
export interface GlobalLegalActionsSummary {
  hasTurnMaterial: boolean;
  hasGlobalPlacementAction: boolean;
  hasPhaseLocalInteractiveMove: boolean;
  hasForcedEliminationAction: boolean;
}

/**
 * True when the specified player still has turn-material in the sense of
 * RR-CANON-R201: at least one controlled stack or at least one ring in hand.
 */
export function hasTurnMaterial(state: GameState, player: number): boolean {
  const playerState = state.players.find((p) => p.playerNumber === player);
  if (!playerState) {
    return false;
  }

  if (playerState.ringsInHand > 0) {
    return true;
  }

  for (const stack of state.board.stacks.values()) {
    if (stack.controllingPlayer === player && stack.stackHeight > 0) {
      return true;
    }
  }

  return false;
}

/**
 * True when at least one legal ring placement exists for `player` in `state`,
 * respecting:
 *
 * - own-colour supply caps (ringsPerPlayer),
 * - ringsInHand,
 * - board geometry and collapsed spaces, and
 * - the no-dead-placement rule via validatePlacementOnBoard.
 *
 * This helper is **phase-agnostic**: it ignores state.currentPhase and can be
 * used from movement/territory/victory code as required by R200.
 */
export function hasGlobalPlacementAction(state: GameState, player: number): boolean {
  const positions = enumeratePlacementPositions(state, player);
  return positions.length > 0;
}

/**
 * Internal helper: build a MovementBoardView over the current GameState so we
 * can reuse hasAnyLegalMoveOrCaptureFromOnBoard without duplicating geometry.
 */
function createMovementBoardView(state: GameState): MovementBoardView {
  const board = state.board;
  const boardType = board.type as BoardType;
  const size = board.size;

  return {
    isValidPosition: (pos: Position) => isValidPosition(pos, boardType, size),
    isCollapsedSpace: (pos: Position) => board.collapsedSpaces.has(positionToString(pos)),
    getStackAt: (pos: Position) => {
      const key = positionToString(pos);
      const stack = board.stacks.get(key);
      if (!stack) {
        return undefined;
      }
      return {
        controllingPlayer: stack.controllingPlayer,
        capHeight: stack.capHeight,
        stackHeight: stack.stackHeight,
      };
    },
    getMarkerOwner: (pos: Position) => {
      const key = positionToString(pos);
      const marker = board.markers.get(key);
      return marker?.player;
    },
  };
}

/**
 * Internal helper: true when **any** non-capture movement or overtaking
 * capture exists for `player` somewhere on the board, irrespective of
 * state.currentPhase. This is used when evaluating forced-elimination
 * preconditions (R072/R100/R205).
 */
function hasAnyGlobalMovementOrCapture(state: GameState, player: number): boolean {
  const boardType = state.board.type as BoardType;
  const view = createMovementBoardView(state);

  for (const stack of state.board.stacks.values()) {
    if (stack.controllingPlayer !== player || stack.stackHeight <= 0) {
      continue;
    }

    if (hasAnyLegalMoveOrCaptureFromOnBoard(boardType, stack.position, player, view)) {
      return true;
    }
  }

  return false;
}

/**
 * True when the **active** player has at least one phase-local interactive
 * move in the current phase (R200, R204):
 *
 * - ring_placement:
 *   - place_ring anywhere that validatePlacementOnBoard accepts, or
 *   - skip_placement when the aggregate reports it as legal.
 * - movement / capture / chain_capture:
 *   - at least one legal non-capturing move or overtaking capture from any
 *     controlled stack.
 * - line_processing:
 *   - at least one process_line or choose_line_reward decision.
 * - territory_processing:
 *   - at least one process_territory_region or eliminate_rings_from_stack
 *     decision.
 *
 * For non-active players this helper always returns false.
 */
export function hasPhaseLocalInteractiveMove(state: GameState, player: number): boolean {
  if (state.gameStatus !== 'active') {
    return false;
  }
  if (player !== state.currentPlayer) {
    return false;
  }

  const phase: GamePhase = state.currentPhase;

  switch (phase) {
    case 'ring_placement': {
      if (hasGlobalPlacementAction(state, player)) {
        return true;
      }
      const eligibility: SkipPlacementEligibilityResult = evaluateSkipPlacementEligibility(
        state,
        player
      );
      return eligibility.eligible;
    }

    case 'movement':
    case 'capture':
    case 'chain_capture': {
      return hasAnyGlobalMovementOrCapture(state, player);
    }

    case 'line_processing': {
      const moves = enumerateProcessLineMoves(state, player, {
        detectionMode: 'detect_now',
      });
      return moves.length > 0;
    }

    case 'territory_processing': {
      const regionMoves = enumerateProcessTerritoryRegionMoves(state, player);
      const elimMoves = enumerateTerritoryEliminationMoves(state, player);
      return regionMoves.length > 0 || elimMoves.length > 0;
    }

    default:
      return false;
  }
}

/**
 * True when RR-CANON forced-elimination preconditions hold for `player` in
 * `state` (R072, R100, R205):
 *
 * - Game is ACTIVE.
 * - Player controls at least one stack (has material on board).
 * - Player has **no** legal placements (hasGlobalPlacementAction === false).
 * - Player has **no** legal non-capture movement or overtaking capture from
 *   any controlled stack (hasAnyGlobalMovementOrCapture === false).
 *
 * This predicate is intentionally phase-agnostic and may be used by hosts at
 * start-of-turn, during decision-phase exits, or by invariants/soaks.
 * It does **not** perform the elimination; hosts should call a shared
 * elimination helper to actually remove the chosen cap.
 */
export function hasForcedEliminationAction(state: GameState, player: number): boolean {
  if (state.gameStatus !== 'active') {
    return false;
  }

  const playerState = state.players.find((p) => p.playerNumber === player);
  if (!playerState) {
    return false;
  }

  // Must control at least one stack on the board.
  let hasStacks = false;
  for (const stack of state.board.stacks.values()) {
    if (stack.controllingPlayer === player && stack.stackHeight > 0) {
      hasStacks = true;
      break;
    }
  }
  if (!hasStacks) {
    return false;
  }

  const anyPlacement = hasGlobalPlacementAction(state, player);
  const anyMoveOrCapture = hasAnyGlobalMovementOrCapture(state, player);

  return !anyPlacement && !anyMoveOrCapture;
}

/**
 * A single stack that is eligible for forced elimination.
 */
export interface ForcedEliminationOption {
  /** Position of the stack. */
  position: Position;
  /** Height of the current player's cap on this stack. */
  capHeight: number;
  /** Total stack height. */
  stackHeight: number;
  /** Move ID for the corresponding eliminate_rings_from_stack move. */
  moveId: string;
}

/**
 * Enumerate all stacks that are eligible for forced elimination for a player.
 *
 * When a player is blocked (has stacks but no legal placement, movement, or
 * capture actions), they must eliminate a cap from one of their stacks. This
 * function returns all eligible stacks so hosts can present a choice to the
 * player (for human players) or auto-select (for AI).
 *
 * Returns an empty array if the player is not in a forced elimination state.
 */
export function enumerateForcedEliminationOptions(
  state: GameState,
  player: number
): ForcedEliminationOption[] {
  if (!hasForcedEliminationAction(state, player)) {
    return [];
  }

  const options: ForcedEliminationOption[] = [];

  for (const stack of state.board.stacks.values()) {
    if (stack.controllingPlayer !== player || stack.stackHeight <= 0) {
      continue;
    }
    const capHeight: number = typeof stack.capHeight === 'number' ? stack.capHeight : 0;

    options.push({
      position: stack.position,
      capHeight,
      stackHeight: stack.stackHeight,
      moveId: `forced-elim-${positionToString(stack.position)}`,
    });
  }

  return options;
}

/**
 * Canonical host-level forced-elimination operator aligned with RR-CANON-R205.
 *
 * This helper is intentionally small and purely functional so that:
 * - turn/phase orchestrators (shared or host-specific) can invoke a single
 *   implementation when a player is blocked with stacks (R072/R100/R205), and
 * - invariant/soak harnesses can apply or simulate forced elimination without
 *   re-deriving stack-selection or elimination bookkeeping.
 *
 * Preconditions (mirroring hasForcedEliminationAction):
 * - state.gameStatus === 'active'
 * - player controls at least one stack on the board
 * - player has no legal placements and no legal movement/capture actions
 *
 * The implementation:
 * - If targetPosition is provided, uses that stack for elimination.
 * - Otherwise, selects a stack controlled by `player`, preferring the smallest
 *   positive capHeight and falling back to the first stack when no caps exist.
 * - Eliminates that stack's cap via applyEliminateRingsFromStackDecision; and
 * - Returns the updated GameState along with basic accounting metadata so
 *   callers can assert INV-ELIMINATION-MONOTONIC / INV-S-MONOTONIC.
 *
 * @param state - Current game state
 * @param player - Player who must eliminate
 * @param targetPosition - Optional: specific stack position chosen by player.
 *                         If not provided, auto-selects (smallest cap heuristic).
 */
export interface ForcedEliminationOutcome {
  /** Updated state after applying forced elimination. */
  nextState: GameState;
  /** Player who paid the elimination cost (always `player`). */
  eliminatedPlayer: number;
  /** Position of the stack that was chosen for elimination, if any. */
  eliminatedFrom?: Position;
  /** Number of rings eliminated from the chosen stack. */
  eliminatedCount: number;
}

export function applyForcedEliminationForPlayer(
  state: GameState,
  player: number,
  targetPosition?: Position
): ForcedEliminationOutcome | undefined {
  if (!hasForcedEliminationAction(state, player)) {
    return undefined;
  }

  let chosenStack:
    | {
        position: Position;
        capHeight: number;
        stackHeight: number;
      }
    | undefined;

  // If a specific target position was provided (player made a choice), use it
  if (targetPosition) {
    const targetKey = positionToString(targetPosition);
    const stack = state.board.stacks.get(targetKey);
    if (stack && stack.controllingPlayer === player && stack.stackHeight > 0) {
      const capHeight: number = typeof stack.capHeight === 'number' ? stack.capHeight : 0;
      chosenStack = {
        position: stack.position,
        capHeight,
        stackHeight: stack.stackHeight,
      };
    }
  }

  // Auto-select if no target provided or target was invalid
  if (!chosenStack) {
    let smallestCap = Number.POSITIVE_INFINITY;

    for (const stack of state.board.stacks.values()) {
      if (stack.controllingPlayer !== player || stack.stackHeight <= 0) {
        continue;
      }
      const capHeight: number = typeof stack.capHeight === 'number' ? stack.capHeight : 0;

      if (capHeight > 0 && capHeight < smallestCap) {
        smallestCap = capHeight;
        chosenStack = {
          position: stack.position,
          capHeight,
          stackHeight: stack.stackHeight,
        };
      } else if (!chosenStack) {
        // Fallback: remember the first stack so that we always have a target
        // even when all caps are zero in legacy or degenerate states.
        chosenStack = {
          position: stack.position,
          capHeight,
          stackHeight: stack.stackHeight,
        };
      }
    }
  }

  if (!chosenStack) {
    return undefined;
  }

  const move: Move = {
    id: `forced-elim-${positionToString(chosenStack.position)}`,
    type: 'eliminate_rings_from_stack',
    player,
    to: chosenStack.position,
    eliminatedRings: [{ player, count: Math.max(1, chosenStack.capHeight || 0) }],
    eliminationFromStack: {
      position: chosenStack.position,
      capHeight: chosenStack.capHeight,
      totalHeight: chosenStack.stackHeight,
    },
    // Deterministic placeholders – callers should not rely on these fields.
    timestamp: new Date(0),
    thinkTime: 0,
    moveNumber: 0,
  } as Move;

  const beforeTotal =
    (state as GameState & { totalRingsEliminated?: number }).totalRingsEliminated ?? 0;

  const { nextState } = applyEliminateRingsFromStackDecision(state, move);

  const afterTotal =
    (nextState as GameState & { totalRingsEliminated?: number }).totalRingsEliminated ?? 0;

  const eliminatedDelta =
    afterTotal > beforeTotal ? afterTotal - beforeTotal : Math.max(1, chosenStack.capHeight || 0);

  return {
    nextState,
    eliminatedPlayer: player,
    eliminatedFrom: chosenStack.position,
    eliminatedCount: eliminatedDelta,
  };
}

/**
 * Compute the R200 global legal action summary for a given player.
 */
export function computeGlobalLegalActionsSummary(
  state: GameState,
  player: number
): GlobalLegalActionsSummary {
  const material = hasTurnMaterial(state, player);
  const placements = hasGlobalPlacementAction(state, player);
  const phaseMoves = hasPhaseLocalInteractiveMove(state, player);
  const forcedElim = hasForcedEliminationAction(state, player);

  return {
    hasTurnMaterial: material,
    hasGlobalPlacementAction: placements,
    hasPhaseLocalInteractiveMove: phaseMoves,
    hasForcedEliminationAction: forcedElim,
  };
}

/**
 * ANM(state) predicate for the active player, mirroring RR-CANON-R202/R203:
 *
 * ANM(state) holds iff:
 * - state.gameStatus === 'active', and
 * - currentPlayer has turn-material (R201), and
 * - G(state, currentPlayer) is empty:
 *   - no global placements,
 *   - no phase-local interactive moves, and
 *   - no forced-elimination action available.
 *
 * Hosts and invariant harnesses can use this helper to enforce
 * INV-ACTIVE-NO-MOVES and INV-ANM-TURN-MATERIAL-SKIP.
 */
export function isANMState(state: GameState): boolean {
  if (state.gameStatus !== 'active') {
    return false;
  }

  const player = state.currentPlayer;
  if (!hasTurnMaterial(state, player)) {
    return false;
  }

  const summary = computeGlobalLegalActionsSummary(state, player);

  return (
    !summary.hasGlobalPlacementAction &&
    !summary.hasPhaseLocalInteractiveMove &&
    !summary.hasForcedEliminationAction
  );
}

/**
 * Convenience wrappers for S/T-style progress metrics used by the
 * invariants framework. S matches computeProgressSnapshot(state).S; T is
 * currently defined as "collapsed + eliminated" so that:
 *
 * - S(state) = markers + collapsed + eliminated (INV-S-MONOTONIC), and
 * - T(state) = collapsed + eliminated (useful for decision-phase chains
 *   that may leave markers unchanged).
 *
 * These helpers are intentionally small so that invariant tests and soaks
 * do not have to re-derive metric definitions.
 */
export function computeSMetric(state: GameState): number {
  return computeProgressSnapshot(state).S;
}

export function computeTMetric(state: GameState): number {
  const snapshot = computeProgressSnapshot(state);
  return snapshot.collapsed + snapshot.eliminated;
}

File: ai-service/app/rules/global_actions.py
from __future__ import annotations

from dataclasses import dataclass

from app.game_engine import GameEngine
from app.models import GamePhase, GameState, GameStatus, MoveType, Position
from app.rules.elimination import EliminationContext, has_eligible_elimination_target


def _did_current_turn_include_recovery_slide(state: GameState, player: int) -> bool:
    """Check if current turn includes a RECOVERY_SLIDE move.

    Used to derive territory elimination context (RR-CANON-R114).
    Mirrors TS didCurrentTurnIncludeRecoverySlide in TerritoryAggregate.ts.
    """
    for move in reversed(state.move_history):
        if move.player != player:
            break
        if move.type == MoveType.RECOVERY_SLIDE:
            return True
    return False


def _derive_territory_elimination_context(state: GameState, player: int) -> EliminationContext:
    """Derive elimination context for territory processing.

    Mirrors TS deriveTerritoryEliminationContext in TerritoryAggregate.ts.

    Args:
        state: Current game state
        player: Player to check

    Returns:
        RECOVERY if turn included recovery slide, TERRITORY otherwise
    """
    if _did_current_turn_include_recovery_slide(state, player):
        return EliminationContext.RECOVERY
    return EliminationContext.TERRITORY

"""
Global-actions and ANM helpers for the Python rules engine.

This mirrors src/shared/engine/globalActions.ts and RR-CANON R200–R207
for use in invariants, soaks, and higher-level diagnostics.
"""


@dataclass(frozen=True)
class GlobalLegalActionsSummary:
    """Summary of the R200 global action surface G(state, player).

    Fields:
        has_turn_material:
            Player has turn-material in the sense of RR-CANON-R201
            (at least one controlled stack or a ring in hand).
        has_global_placement_action:
            At least one legal ring placement exists for the player,
            ignoring state.current_phase (RR-CANON-R200, R080–R082).
        has_phase_local_interactive_move:
            At least one phase-local interactive move exists for the
            player in state.current_phase (R200, R204).
        has_forced_elimination_action:
            Host-level forced elimination is available for the player
            under RR-CANON-R072/R100/R205.
    """

    has_turn_material: bool
    has_global_placement_action: bool
    has_phase_local_interactive_move: bool
    has_forced_elimination_action: bool


@dataclass(frozen=True)
class ForcedEliminationOutcome:
    """Result of applying host-level forced elimination for a player."""

    next_state: GameState
    eliminated_player: int
    eliminated_from: Position | None
    eliminated_count: int


def has_turn_material(state: GameState, player: int) -> bool:
    """Return True if player has turn-material in state (RR-CANON-R201)."""

    player_state = next(
        (p for p in state.players if p.player_number == player),
        None,
    )
    if player_state is None:
        return False

    if player_state.rings_in_hand > 0:
        return True

    return any(stack.controlling_player == player and stack.stack_height > 0 for stack in state.board.stacks.values())


def has_global_placement_action(state: GameState, player: int) -> bool:
    """Return True if any legal ring placement exists for ``player``.

    This is a boolean predicate used by ANM/global-action checks.

    Important: do **not** enumerate all placement moves here.

    ``GameEngine._get_ring_placement_moves`` is intentionally expensive because
    it fully enumerates placements and runs no-dead-placement validation for
    each candidate. Calling that on every post-move ANM check can turn
    self-play and fitness evaluation into an accidental quadratic/cubic loop.

    ``GameEngine._has_any_valid_placement_fast`` is semantically equivalent for
    existence checks (same caps + no-dead-placement), but returns early.
    """

    return GameEngine._has_any_valid_placement_fast(state, player)


def has_phase_local_interactive_move(
    state: GameState,
    player: int,
) -> bool:
    """Return True if player has an interactive move in the current phase.

    Interactive moves exclude host-synthesised ``no_*_action`` bookkeeping
    moves. This mirrors the TS helper in
    ``src/shared/engine/globalActions.ts``:

    - ``ring_placement``:
        ``place_ring`` or ``skip_placement``.
    - ``movement`` / ``capture`` / ``chain_capture``:
        movement, capture, or ``recovery_slide`` moves.
    - ``line_processing``:
        ``process_line`` or ``choose_line_option``.
    - ``territory_processing``:
        ``choose_territory_option``, ``eliminate_rings_from_stack``, or
        ``skip_territory_processing``.
    - ``forced_elimination``:
        availability of forced-elimination options for the player.

    For non-active games or non-current players this helper always returns
    False.
    """

    if state.game_status != GameStatus.ACTIVE:
        return False

    if player != state.current_player:
        return False

    phase = state.current_phase

    # Forced elimination is its own phase; treat the existence of any FE option
    # as an interactive choice for this predicate.
    if phase == GamePhase.FORCED_ELIMINATION:
        return has_forced_elimination_action(state, player)

    # LINE_PROCESSING: Per RR-CANON-R123, when pending_line_reward_elimination
    # is True, the player must execute an eliminate_rings_from_stack move.
    # This counts as an interactive move for ANM calculation, matching TS
    # hasPhaseLocalInteractiveMove behavior in globalActions.ts:254-267.
    if phase == GamePhase.LINE_PROCESSING:
        if state.pending_line_reward_elimination:
            # Player has controlled stacks -> elimination moves exist
            for stack in state.board.stacks.values():
                if stack.controlling_player == player and stack.stack_height > 0:
                    return True
            # No controlled stacks means no elimination possible (edge case)
            return False

        # Otherwise check for line processing moves using fresh line detection.
        # Per RR-CANON-R204, this uses fresh line detection to determine if any
        # lines exist for the player.
        moves = GameEngine._get_line_processing_moves(state, player)
        return len(moves) > 0

    # TERRITORY_PROCESSING: Per RR-CANON-R145, when the last move by this player
    # was a choose_territory_option (with region data), there's a pending
    # self-elimination that counts as an interactive move. This mirrors TS
    # globalActions.ts which calls BOTH enumerateProcessTerritoryRegionMoves AND
    # enumerateTerritoryEliminationMoves for the ANM check.
    #
    # HEX-PARITY-03 fix: Check for pending territory self-elimination explicitly,
    # similar to the pending_line_reward_elimination check in line_processing.
    if phase == GamePhase.TERRITORY_PROCESSING:
        # Check for pending territory self-elimination (RR-CANON-R145)
        # This matches TS getPendingTerritorySelfEliminationRegion logic
        last_move = state.move_history[-1] if state.move_history else None
        if last_move is not None and last_move.player == player:
            if last_move.type in (MoveType.CHOOSE_TERRITORY_OPTION, MoveType.PROCESS_TERRITORY_REGION):
                # Check if the move has region data (required for pending elimination)
                regions = getattr(last_move, "disconnected_regions", None)
                if regions:
                    # Pending self-elimination: check for eligible elimination targets
                    # outside the processed region using canonical eligibility check.
                    # HEX-PARITY-02 FIX: Derive elimination context like TypeScript does.
                    # If turn included recovery slide, use RECOVERY context (checks for
                    # buried rings). Otherwise use TERRITORY context (checks for control).
                    # This matches TS deriveTerritoryEliminationContext.
                    processed_region_keys = {p.to_key() for p in regions[0].spaces}
                    elimination_context = _derive_territory_elimination_context(state, player)

                    # Build stacks dict for eligibility check
                    stacks_dict = {}
                    for stack in state.board.stacks.values():
                        pos_key = stack.position.to_key()
                        stacks_dict[pos_key] = {
                            "rings": list(stack.rings),
                            "controlling_player": stack.controlling_player,
                        }

                    return has_eligible_elimination_target(
                        stacks_dict,
                        player,
                        elimination_context,
                        exclude_positions=processed_region_keys,
                    )

        # No pending elimination: check for region processing moves
        moves = GameEngine._get_territory_processing_moves(state, player)
        return len(moves) > 0

    # For other phases, consult the canonical move generator. It returns only
    # interactive moves; bookkeeping no_* moves are surfaced via
    # get_phase_requirement / synthesize_bookkeeping_move instead.
    moves = GameEngine.get_valid_moves(state, player)
    if not moves:
        return False

    if phase == GamePhase.RING_PLACEMENT:
        interactive_types = {
            MoveType.PLACE_RING,
            MoveType.SKIP_PLACEMENT,
        }
    elif phase == GamePhase.MOVEMENT:
        interactive_types = {
            MoveType.MOVE_STACK,
            MoveType.MOVE_RING,
            MoveType.BUILD_STACK,
            MoveType.OVERTAKING_CAPTURE,
            MoveType.CONTINUE_CAPTURE_SEGMENT,
            MoveType.CHAIN_CAPTURE,
            MoveType.RECOVERY_SLIDE,
        }
    elif phase in (GamePhase.CAPTURE, GamePhase.CHAIN_CAPTURE):
        interactive_types = {
            MoveType.OVERTAKING_CAPTURE,
            MoveType.CONTINUE_CAPTURE_SEGMENT,
        }
    else:
        return False

    return any(move.type in interactive_types for move in moves)


def has_forced_elimination_action(state: GameState, player: int) -> bool:
    """Return True if host-level forced elimination is available for player."""

    if state.game_status != GameStatus.ACTIVE:
        return False

    moves = GameEngine._get_forced_elimination_moves(state, player)
    return bool(moves)


def global_legal_actions_summary(
    state: GameState,
    player: int,
) -> GlobalLegalActionsSummary:
    """Compute the R200 global legal action summary for the player."""

    material = has_turn_material(state, player)
    placements = has_global_placement_action(state, player)
    phase_moves = has_phase_local_interactive_move(state, player)
    forced = has_forced_elimination_action(state, player)

    return GlobalLegalActionsSummary(
        has_turn_material=material,
        has_global_placement_action=placements,
        has_phase_local_interactive_move=phase_moves,
        has_forced_elimination_action=forced,
    )


def is_anm_state(state: GameState) -> bool:
    """Return True iff ANM(state, currentPlayer) holds (RR-CANON-R202)."""

    if state.game_status != GameStatus.ACTIVE:
        return False

    player = state.current_player
    if not has_turn_material(state, player):
        return False

    summary = global_legal_actions_summary(state, player)

    return (
        not summary.has_global_placement_action
        and not summary.has_phase_local_interactive_move
        and not summary.has_forced_elimination_action
    )


def apply_forced_elimination_for_player(
    state: GameState,
    player: int,
) -> ForcedEliminationOutcome | None:
    """Apply one host-level forced elimination for ``player``, if legal.

    The function mutates ``state`` in-place and returns a
    ForcedEliminationOutcome when an elimination occurs, or None when
    RR-CANON-R072/R100/R205 preconditions are not satisfied.
    """

    if not has_forced_elimination_action(state, player):
        return None

    moves = GameEngine._get_forced_elimination_moves(state, player)
    if not moves:
        return None

    fe_move = moves[0]

    before_total = state.total_rings_eliminated
    GameEngine._apply_forced_elimination(state, fe_move)
    after_total = state.total_rings_eliminated

    eliminated_delta = after_total - before_total
    if eliminated_delta <= 0:
        eliminated_delta = 1

    return ForcedEliminationOutcome(
        next_state=state,
        eliminated_player=player,
        eliminated_from=fe_move.to,
        eliminated_count=eliminated_delta,
    )
